backend/src/main/java/ai/verta/modeldb/experimentRun/FutureExperimentRunDAO.java

```diff
@@ -409,6 +414,82 @@ public FutureExperimentRunDAO(
         .thenCompose(unused -> deleteExperimentRuns(runIds), executor);
   }
 
+  /**
+   * @param newDatasets : new datasets for privilege check
+   * @param errorOut : Throw error while creation (true) otherwise we will keep it silent (false)
+   * @return {@link List} : accessible datasets
+   */
+  private InternalFuture<List<Artifact>> filterAndGetPrivilegedDatasetsOnly(
+      List<Artifact> newDatasets, boolean errorOut) {
+    List<String> newDatasetLinkedIds = new ArrayList<>();
+    for (Artifact dataset : newDatasets) {
+      String datasetVersionId = dataset.getLinkedArtifactId();
+      if (!datasetVersionId.isEmpty()) {
+        newDatasetLinkedIds.add(datasetVersionId);
+      }
+    }
+
+    if (newDatasetLinkedIds.isEmpty()) {
+      return InternalFuture.completedInternalFuture(newDatasets);
+    }
+    return jdbi.withHandle(
+            handle -> {
+              Map<String, Long> datasetVersionDatasetMap = new HashMap<>();
+              handle
+                  .createQuery(
+                      " SELECT commit_hash, repository_id FROM repository_commit WHERE commit_hash IN (<datasetIds>) ")
+                  .bindList("datasetIds", newDatasetLinkedIds)
+                  .map(
+                      (rs, ctx) ->
+                          datasetVersionDatasetMap.put(
+                              rs.getString("commit_hash"), rs.getLong("repository_id")))
+                  .list();
+              return datasetVersionDatasetMap;
+            })
+        .thenCompose(
+            datasetVersionDatasetMap -> {
+              List<String> accessibleDatasetVersionIds = new ArrayList<>();
+              List<InternalFuture<Artifact>> internalFutures = new LinkedList<>();
+              for (Artifact dataset : newDatasets) {
+                String datasetVersionId = dataset.getLinkedArtifactId();
+                if (!datasetVersionId.isEmpty()
+                    && !accessibleDatasetVersionIds.contains(datasetVersionId)) {
```

Comment:
You're building your own cache of ids already processed, which introduces the race condition below. This is classical serial thinking, which also makes the code more confusing. Instead, here's my suggestion of low:
- create a set out of the ids of `newDatasets`. Now you have de-dup and can drop `accessibleDatasetVersionIds` completely
- for each item in the set, you 1) filter based on `datasetVersionDatasetMap`, 2) map it to `checkEntityPermissionBasedOnResourceTypes` (you'll have to transform List<Future> to Future<List>), 3) filter only the successes, 4) convert to a set
- Now you have a set of valid ids, then you can filter `newDatasets` based on whether the id is present in the set or not

Reply:
Changed entire code as per this suggestions